Refine suggestions about missing components

rami3l opened the pull request:
Closes #3418. I also cleaned up the current code a bit along the way.

cc #2355

rbtcollins commented on src/errors.rs:
```diff
@@ -171,57 +171,69 @@ fn remove_component_msg(cs: &Component, manifest: &Manifest, toolchain: &str) ->
-        if toolchain.starts_with("nightly") {
-            let _ = write!(
+    match cs {
+        [] => panic!("`component_unavailable_msg` should not be called with an empty collection of unavailable components"),
```
This has added a new panic to code that couldn't panic before. For `rustup` panics are quite a concern, though in an error path is -perhaps- acceptable.

In this case I can see there is only one construction point for the error value that formats with this helper. Please add a doc comment to the variant that the helper will panic; we can look at using https://docs.rs/nonempty/latest/nonempty/ or similar later.

rami3l replied:
@rbtcollins IMHO it didn't. This line corresponds to the original `assert!(!cs.is_empty());` line which is also a panic point if `cs` matches `[]`.
... but I'd be happy to change this for the better by, for example, adding a "panic" section in the docs.

rbtcollins replied:
ah thats fair! I think we should still add a doccomment to the variant, because this is a timebomb otherwise.

rbtcollins commented on src/errors.rs:
```diff
@@ -171,57 +171,69 @@ fn remove_component_msg(cs: &Component, manifest: &Manifest, toolchain: &str) ->
+                    "If you are trying to update an existing toolchain and no longer need these components,\n\
+                    you might want to remove them first with:\n\n{remove_msg}\n\n{TOOLSTATE_MSG}",
+            );
+        }
```
its quite hard to determine what you've actually changed here - the change combines a refactoring (fine), and new message  logic (also fine), but even though I've looked in:
- the PR message
- the commit messages
- test changes (none)
I can't tell what the actual intent is without manually stepping through in my head every possible case.

It would be a lot easier to reason about if the refactoring and the functional changes were separate (e.g. two commits in one PR - like you did for the other module that you refactored).

I realise that that isn't always possible, and if it wasn't here then I guess its as good as we can get :).

The other norm we have is to ensure test coverage for functional behaviours. We're in the process of moving tests closer to the domain for the test, so looking at the existing test coverage isn't necessarily the best way to figure out where tests should go - but its ok to put them in the blackbox test suite under `tests/` and we can move them later.

rami3l replied:
> It would be a lot easier to reason about if the refactoring and the functional changes were separate (e.g. two commits in one PR - like you did for the other module that you refactored).

Yes, I am aware of that. I will leave the current commit as it is, but I will definitely improve this from now on.

> The other norm we have is to ensure test coverage for functional behaviours. We're in the process of moving tests closer to the domain for the test, so looking at the existing test coverage isn't necessarily the best way to figure out where tests should go - but its ok to put them in the blackbox test suite under `tests/` and we can move them later.

I'd be glad to add blackbox tests for this one (which will be done in another commit), please bear with me 🙏

rami3l replied:
@rbtcollins The functional test changes now reflect this PR's original intention.

rami3l commented:
@rbtcollins Looking at the blackbox test suite, it seems to me that this fix is more than what it seems to be.

In toolchain **updates**, `component_unavailable_msg`  

https://github.com/rust-lang/rustup/blob/a4dd7d03501def7ee3e4e2961bca7f6ec3b2e1aa/src/errors.rs#L173

does get called, but the generated error then gets **caught and re-thrown**

https://github.com/rust-lang/rustup/blob/a4dd7d03501def7ee3e4e2961bca7f6ec3b2e1aa/src/dist/dist.rs#L938-L946

to `components_missing_msg`

https://github.com/rust-lang/rustup/blob/a4dd7d03501def7ee3e4e2961bca7f6ec3b2e1aa/src/dist/dist.rs#L45

, which instead suggests using a `minimal` profile, so the original suggestion in the former (uninstall the component in question) might not be applicable at all anymore, as the error message from the former can only be seen in some edge cases (like adding `miri` to `stable`, but not upgrading to a new `nightly` which lacks `rls`).

To move this fix forward, I could simply remove the suggestion in `component_unavailable_msg`, however, I want to know more about the semantic differences between these two functions in order to make the final decision.

For example, since this help message is very well-written, I wish to be able to see it in toolchain updates as well:

https://github.com/rust-lang/rustup/blob/a4dd7d03501def7ee3e4e2961bca7f6ec3b2e1aa/src/errors.rs#L22-L28

rami3l commented:
Note: `dist::manifestation::tests::unavailable_component` and friends test `rustup::dist::manifestation::Manifestation::update` directly (by checking the very `rustup component remove` suggestion), which happens before the catch-rethrow dance in `rustup::dist::dist::try_update_from_dist_`, so these tests are expected to fail at the moment.

... and they are not accurate anyway since the string form of those errors (before the catch-rethrow dance) is NOT ALWAYS visible to end users, which might cause further confusions for maintainers later on.

rami3l commented:
I've yet to see any use case or test case that disobeys the observation stated in https://github.com/rust-lang/rustup/pull/3453/commits/6b5cb9f904ce9acdacc416848f5ba050a0c6849f 's commit message.

Therefore, I think by now it is reasonable to do the following:

- [x] Add a new blackbox test case to ensure that #3418 does get fixed.
- [x] Move `TOOLSTATE_MSG` to `dist` to serve toolchain-wide operations (since it also talks about toolchain changes) by appending it to the end of `component_missing_msg`'s output.

rami3l commented on src/dist/manifestation/tests.rs:
```diff
@@ -992,13 +1019,24 @@ fn unavailable_components_is_target() {
+                        .iter()
+                        .map(|c| c.description(&manifest))
+                        .collect::<Vec<_>>();
+                    assert_eq!(
+                        descriptions,
+                        [
+                            "'rust-std' for target 'x86_64-apple-darwin'",
+                            "'rust-std' for target 'i686-apple-darwin'",
+                            "'rust-std' for target 'i686-unknown-linux-gnu'"
+                        ]
+                    );
```
Here, I am changing the tests so that they no longer rely on the textual representation of the error.

This has actually exposed a mistake in the original test case, which only tested the INCLUSION rather than the whole collection.

rbtcollins commented:
> I've yet to see any use case or test case that disobeys the observation stated in [6b5cb9f](https://github.com/rust-lang/rustup/commit/6b5cb9f904ce9acdacc416848f5ba050a0c6849f) 's commit message.
> 
> Therefore, I think by now it is reasonable to do the following:
> 
>     * [x]  Add a new blackbox test case to ensure that [Confusing `component add` error message #3418](https://github.com/rust-lang/rustup/issues/3418) does get fixed.
> 
>     * [x]  Move `TOOLSTATE_MSG` to `dist` to serve toolchain-wide operations (since it also talks about toolchain changes) by appending it to the end of `component_missing_msg`'s output.

seems reasonable to me

rami3l commented:
> > I've yet to see any use case or test case that disobeys the observation stated in [6b5cb9f](https://github.com/rust-lang/rustup/commit/6b5cb9f904ce9acdacc416848f5ba050a0c6849f) 's commit message.
> > Therefore, I think by now it is reasonable to do the following:
> > ```
> > * [x]  Add a new blackbox test case to ensure that [Confusing `component add` error message #3418](https://github.com/rust-lang/rustup/issues/3418) does get fixed.
> > 
> > * [x]  Move `TOOLSTATE_MSG` to `dist` to serve toolchain-wide operations (since it also talks about toolchain changes) by appending it to the end of `component_missing_msg`'s output.
> > ```
> 
> seems reasonable to me

@rbtcollins  ... and those two tasks have been done, together with a new black box test case for the latter.
Would you mind reviewing my changes?

rami3l committed:
Clean up some `manifestation` logic

rami3l committed:
Add test to ensure resolution of #3418

rami3l committed:
Move `TOOLSTATE_MSG` to `dist` to serve toolchain-wide operations

rami3l committed:
Delete suggestions of removing the relevant component from `component_unavailable_msg`

After digging into the codebase, I realized that the message generated
by `component_unavailable_msg` is used only when:

- Some components are missing from the toolchain so that the
  installation can no longer proceed;
- We are not installing these components as a part of a
  toolchain-wide operation (e.g. updating the existing `nightly`),
  which is covered by `components_missing_msg` by catching the
  `RustupError::RequestedComponentsUnavailable` and re-throwing
  it as a `DistError::ToolchainComponentsMissing`
  (see <https://github.com/rust-lang/rustup/pull/3453#issuecomment-1685269047>
  for more info).

Thus, I decided to remove the `rustup component remove` suggestion
altogether.

djc commented on src/dist/dist.rs:
```diff
@@ -43,6 +43,14 @@ static TOOLCHAIN_CHANNELS: &[&str] = &[
     r"\d{1}\.\d{1,3}(?:\.\d{1,2})?",
 ];
 
+const TOOLSTATE_MSG: &str =
```
The commit message mentions to `Move` this, but I don't see it being removed anywhere. Is that an oversight?

rami3l replied:
Sorry for previously being unclear about this. This is an old PR and I've learned a lot on how to separate a big patch into relevant small commits since then.

The message was previously deleted in https://github.com/rust-lang/rustup/pull/3453/commits/9107c0235ceeb1b227e333495903b78acb135499. Maybe it's better to say that I "recycled" it here since this error message has been used in a wrong module for another kind of error (https://github.com/rust-lang/rustup/pull/3453#issuecomment-1685269047).

rami3l replied:
@djc I rewrote the history to make the move actually appear in the last commit. Sorry about the inconvenience as I'm still improving my collaboration skills!

djc commented on src/dist/dist.rs:
```diff
@@ -43,46 +43,57 @@ static TOOLCHAIN_CHANNELS: &[&str] = &[
+            );
+
+            if toolchain.starts_with("nightly") {
+                _ = write!(buf, "{nightly_tips}");
```
Nit: not a fan of this change to remove the `let` keywords. IMO the `let _ =` is more conventional.

rami3l replied:
Will do.

djc reviewed (commented):
Please keep unrelated changes in separate commits, makes it much easier to review.

rami3l committed:
Refactor `components_*_msg`

rami3l committed:
Add `Panics` sections to docstrings
